### DISCUSSION TITLE
Inner working of PyCaret data model

### DISCUSSION THREAD

**Discussion by CarloNicolini** [pycaret/pycaret#1152] [General]
PyCaret is somewhat different from normal ML libraries, in the sense that in only requires some calls to global functions.
I was wondering if it possible to expose some of its objected oriented part.

As a developer and data-scientist, I always ask myself where PyCaret stores the current state of things, for example, after a call to `setup` where are all the internal variables stored that can be called and used after a call to `compare_models()`?

Similarly to the sklearn pipeline how can I expose and compose a PyCaret pipeline using and storing objects?

**Comment by Yard1** [comment]
The currently released versions uses multiple global variables to store the data. You can set and obtain them with `set_config`/`get_config`. Next *major* release will see the architecture move to an object oriented model while retaining the functional API.

**Reply by ergut** [reply]
Do you have any update on the plans for the architectural change? The article "[PyCaret for Classification: An Honest Review](https://towardsdatascience.com/pycaret-review-65cbe2f663bb)" by Bence Komarniczky lists some of the shortcomings of the current architecture.

**Reply by Yard1** [reply]
We are still working on the 3.0 update which will update the architecture.
